### ISSUE TITLE
The channelMax limit is reached with rabbit having higher channel max than client

### ISSUE THREAD

**Issue by ldai100** [spring-projects/spring-amqp#999]
Hi, I've been getting this error for a while and I don't see anything wrong in my configuration. Here's my stack trace:
```
org.springframework.amqp.AmqpResourceNotAvailableException: The channelMax limit is reached. Try later.
	at org.springframework.amqp.rabbit.connection.SimpleConnection.createChannel(SimpleConnection.java:59)
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory$ChannelCachingConnectionProxy.createBareChannel(CachingConnectionFactory.java:1326)
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory$ChannelCachingConnectionProxy.access$200(CachingConnectionFactory.java:1312)
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.doCreateBareChannel(CachingConnectionFactory.java:654)
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.createBareChannel(CachingConnectionFactory.java:637)
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.getCachedChannelProxy(CachingConnectionFactory.java:607)
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.getChannel(CachingConnectionFactory.java:498)
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.access$1600(CachingConnectionFactory.java:101)
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory$ChannelCachingConnectionProxy.createChannel(CachingConnectionFactory.java:1331)
	at org.springframework.amqp.rabbit.core.RabbitTemplate.doExecute(RabbitTemplate.java:2067)
	at org.springframework.amqp.rabbit.core.RabbitTemplate.lambda$execute$13(RabbitTemplate.java:2024)
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:287)
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:180)
	at org.springframework.amqp.rabbit.core.RabbitTemplate.execute(RabbitTemplate.java:2023)
	at org.springframework.amqp.rabbit.core.RabbitTemplate.send(RabbitTemplate.java:982)
	at org.springframework.amqp.rabbit.core.BatchingRabbitTemplate.send(BatchingRabbitTemplate.java:67)
	at org.springframework.amqp.rabbit.core.RabbitTemplate.convertAndSend(RabbitTemplate.java:1088)
	at com.whiteops.sentinel.server.rabbit.RabbitContext.send(RabbitContext.java:291)
	at com.whiteops.sentinel.server.receiver.ReceiverService.receive(ReceiverService.java:185)
	at sun.reflect.GeneratedMethodAccessor73.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:189)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:138)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:102)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:800)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1038)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:942)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1005)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:908)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:660)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:882)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:320)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:74)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:215)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:357)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:270)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:92)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:200)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:200)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:490)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:408)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:834)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1415)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
```
The channel max I set on rabbit side: ```{channel_max,5000}```
The connection factory setting:
```
 // BUG? Setting the channel cache size does not impact the client/server negotiated max channels
        CachingConnectionFactory connectionFactory = new CachingConnectionFactory();
        connectionFactory.setPort(properties.getPort());
        connectionFactory.setAddresses(properties.getAddresses());
        connectionFactory.setUsername(properties.getUsername());
        connectionFactory.setPassword(properties.getPassword());
        connectionFactory.setCacheMode(CachingConnectionFactory.CacheMode.CHANNEL);
        connectionFactory.setPublisherConfirms(true);
        connectionFactory.setPublisherReturns(false);
        connectionFactory.setConnectionNameStrategy(f -> getSource(receiverProperties.getQueue()));
        connectionFactory.setChannelCacheSize(1000);
        connectionFactory.setChannelCheckoutTimeout(1000);
```
I initially didn't have channel checkout timeout set up, I put it in to try it again but same error still exist.
The version I'm currently using is 2.1.3.RELEASE.

Since rabbit has a way higher channel max limit, this error shouldn't occur right?
Please let me know if you might need anything else from me that would help.  Thank you

**Comment by garyrussell** [comment]
The error should not occur unless you have 5000 channels in use. The stack trace alone is insufficient; please provide a small sample app that reproduces the issue.

Also look in the server logs.

**Comment by ldai100** [comment]
thanks for the quick reply.  Here's picture of my rabbit channels all limited to 1000.
<img width="455" alt="Screen Shot 2019-05-15 at 11 57 59 AM" src="https://user-images.githubusercontent.com/15991751/57790472-cf20fc80-7708-11e9-8263-90b385b72893.png">
![Screen Shot 2019-05-15 at 12 00 02 PM](https://user-images.githubusercontent.com/15991751/57790580-0abbc680-7709-11e9-84a8-2c12e4bde6ac.png)

and the rabbit setting
I don't think I am able to provide a sample app at the moment.
Are there workarounds for this issue?  We did retry policy, although it decreases the number of failures but we are still losing data while doing retry.

**Comment by garyrussell** [comment]
? If your channelMax per connection is 1000 and 1000 are in use then, of course, you'll get that exception if you try to create another.

Look on the connection details screen to see that the channel_max is set as expected.

The only way that exception is thrown is when the client says no channels are available...

```java
	@Override
	public Channel createChannel(boolean transactional) {
		try {
			Channel channel = this.delegate.createChannel();
			if (channel == null) {
				throw new AmqpResourceNotAvailableException("The channelMax limit is reached. Try later.");
			}
			if (transactional) {
				// Just created so we want to start the transaction
				channel.txSelect();
			}
			return channel;
		}
		catch (IOException e) {a
			throw RabbitExceptionTranslator.convertRabbitAccessException(e);
		}
	}
```

and

```java
    /**
     * Create a new channel, using an internally allocated channel number.
     * If <a href="https://www.rabbitmq.com/api-guide.html#recovery">automatic connection recovery</a>
     * is enabled, the channel returned by this method will be {@link Recoverable}.
     * <p>
     * Use {@link #openChannel()} if you want to use an {@link Optional} to deal
     * with a {@null} value.
     *
     * @return a new channel descriptor, or null if none is available
     * @throws IOException if an I/O problem is encountered
     */
    Channel createChannel() throws IOException;
````

**Comment by ldai100** [comment]
Yes, we did locate the createChannel method you mentioned above and understands that the error should only be thrown when no resource is available from rabbitMQ.  Our rabbit node was set to channel_max = 5000.  The reason those channels are at 1000 is because we set the client's configuration:
```
       connectionFactory.setChannelCacheSize(1000);
       connectionFactory.setChannelCheckoutTimeout(1000);
```
Through documentation channelCacheSize simply determines the size of the cache and won't limit anything. If channelCheckoutTimeout was not default, it will limit the number of channels.  In this case, it was 1000.  Even when we don't set channelCheckoutTimeout, same error still exists.  We even tried to set the rabbit side with channel_max = 0 (unlimited channels), this error will still be thrown.

**Comment by garyrussell** [comment]
The fact that your connection has reached 1000 channels, means that you actually needed that number of channels at some time. It is very unusual to need so many channels since, except on the consumer side, they are usually in use for only a small amount of time. Usually, only a few 10s of channels are needed for producers.

How are you actually using the channels? Are you using a `RabbitTemplate` (which reliably returns the channel to the cache after each use, by `close()`ing it). If not, are you `close()` ing the channel after each use?

When I run with checkout timeout, I get a different exception:

```java
@SpringBootApplication
public class Rgh999Application {

	public static void main(String[] args) {
		SpringApplication.run(Rgh999Application.class, args);
	}

	@Bean
	public ApplicationRunner runner(CachingConnectionFactory cf) {
		cf.setChannelCacheSize(42);
		cf.setChannelCheckoutTimeout(1000L);
		return args -> {
			cf.getRabbitConnectionFactory().setRequestedChannelMax(42);
			Connection conn = cf.createConnection();
			for (int i = 0; i < 43; i++) {
				conn.createChannel(false);
			}
			System.in.read();
		};
	}

}
```

```
Caused by: org.springframework.amqp.AmqpTimeoutException: No available channels
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.obtainPermits(CachingConnectionFactory.java:521) ~[spring-rabbit-2.2.0.M1.jar:2.2.0.M1]
```

And with the `cf.setChannelCheckoutTimeout(1000L);` commented out...

```
Caused by: org.springframework.amqp.AmqpResourceNotAvailableException: The channelMax limit is reached. Try later.
	at org.springframework.amqp.rabbit.connection.SimpleConnection.createChannel(SimpleConnection.java:59) ~[spring-rabbit-2.2.0.M1.jar:2.2.0.M1]
```

**Comment by ldai100** [comment]
Yes, I am using the template to manage the channels as I believe that's the best way.

**Comment by garyrussell** [comment]
It's very, very, unusual to need 1000s of channels; so something must be wrong.

**Comment by ldai100** [comment]
Yes, I am using the template for channel management which I believe is the best way.

here's how I set up the connection:

```
 // BUG? Setting the channel cache size does not impact the client/server negotiated max channels
        CachingConnectionFactory connectionFactory = new CachingConnectionFactory();
        connectionFactory.setPort(properties.getPort());
        connectionFactory.setAddresses(properties.getAddresses());
        connectionFactory.setUsername(properties.getUsername());
        connectionFactory.setPassword(properties.getPassword());
        connectionFactory.setCacheMode(CachingConnectionFactory.CacheMode.CHANNEL);
        connectionFactory.setPublisherConfirms(true);
        connectionFactory.setPublisherReturns(false);
        connectionFactory.setConnectionNameStrategy(f -> getSource(receiverProperties.getQueue()));
        connectionFactory.setChannelCacheSize(1000);
        connectionFactory.setChannelCheckoutTimeout(1000);
```
Template creation:
```
            RabbitTemplate template = new RabbitTemplate(sentinelConnectionFactory);
            template.setMessageConverter(converter);
            template.setUsePublisherConnection(true);
```
How I send using the template
```
            template.convertAndSend(exchange, getRoutingKey(isSharded), item, this::persistence, result);
```

**Comment by garyrussell** [comment]
If that's all you are doing , it makes no sense that you have 1000s of channels open. The template closes the channel each time. You should have, at the very most the same number of channels as threads that call the template. So something else is going on.

Can you provide a small app that exhibits this behavior?

**Comment by ldai100** [comment]
We are working on a smaller version of our project that exhibits this problem.  Thanks for the help so far.  I'll share it as soon it's available.

**Comment by ldai100** [comment]
[small-sample.zip](https://github.com/spring-projects/spring-amqp/files/3188248/small-sample.zip)
Here's a small sample App that can reproduce the error.  There is also a standalone test in there to run called SpeedMain.java.  The comments provide commands to execute the test.  We used this build in test as well as locust to test and both of them will throw the channelMax error.

**Comment by ldai100** [comment]
In addition, if I do 
```
//connectionFactory.setPublisherConfirms(true);
//connectionFactory.setPublisherReturns(false);
connectionFactory.setSimplePublisherConfirms(true);
```
and remove 
```
// Received successfully
//template.setConfirmCallback(this::callback);
```
I no longer get channel errors.  So it has to be something with the publisher confirmation callback?

**Comment by garyrussell** [comment]
Ugh (Lombok) - it's a real pain for those of us who don't use it :(

**Comment by garyrussell** [comment]
When using confirms, we have to keep the channel out of the cache until the confirm is received. So it's definitely going to cause an increase of in-use channels.

**Comment by ldai100** [comment]
lol, sorry.  figured using free setters and getters making using spring easier.  Let's say the channels are out of cache and we need more channels, it will keep creating new channels until 1) channel max set by rabbitMQ is reached  or 2) we ran out of resources locally, 3) locally set max is reached and it start blocking.  Am I correct?  In this case, rabbit has higher channel max set up, local channel limit was intentionally set lower than rabbit server, so I the channelMax error shouldn't occur?

**Comment by garyrussell** [comment]
Actually, it's only a pain in my IDE, so it's not so bad (Lombok) given you have a runtime test.

The channel checkout won't block unless you set the channelCheckoutTimeout. The "local max" is simply a max applied to the connection; when it is reached; the client returns null.

**Comment by garyrussell** [comment]
Here's one workaround - just uses a single channel; obviously you'd need some mechanism to exit the while loop during shutdown...

```diff
$ diff small-sample3/src/main/java/com/foo/server/RabbitContext.java small-sample3-modified/src/main/java/com/foo/server/RabbitContext.java 
4a5,11
> 
> import java.util.concurrent.BlockingQueue;
> import java.util.concurrent.ExecutorService;
> import java.util.concurrent.Executors;
> import java.util.concurrent.LinkedBlockingQueue;
> import java.util.concurrent.TimeUnit;
> 
6a14
> 
63a72,73
>     private final BlockingQueue<ItemAndResult<T>> queue = new LinkedBlockingQueue<>();
> 
80a91,109
>         ExecutorService exec = Executors.newSingleThreadExecutor();
>         exec.execute(() -> {
> 	        this.template.invoke(t -> {
> 	        	while(true) {
> 	        		try {
> 						ItemAndResult<T> next = this.queue.poll(5, TimeUnit.SECONDS);
> 						if (next != null) {
> 							t.convertAndSend(exchange, getRoutingKey(), next.getItem(), this::persistence,
> 									next.getResult());
> 						}
> 					}
> 					catch (InterruptedException e) {
> 						Thread.currentThread().interrupt();
> 						break;
> 					}
> 				}
> 				return null;
> 			});
>        	});
128c157
<             template.convertAndSend(exchange, getRoutingKey(), item, this::persistence, result);
---
>             this.queue.put(new ItemAndResult<>(item, result));
188a218,242
>     public static class ItemAndResult<T> {
> 
>     	private final T item;
> 
>     	private SendResult<T> result;
> 
> 		public ItemAndResult(T item, SendResult<T> result) {
> 			this.item = item;
> 			this.result = result;
> 		}
> 
> 		public SendResult<T> getResult() {
> 			return this.result;
> 		}
> 
> 		public void setResult(SendResult<T> result) {
> 			this.result = result;
> 		}
> 
> 		public T getItem() {
> 			return this.item;
> 		}
> 
>     }
> 
```

**Comment by ldai100** [comment]
> Actually, it's only a pain in my IDE, so it's not so bad (Lombok) given you have a runtime test.
> 
> The channel checkout won't block unless you set the channelCheckoutTimeout. The "local max" is simply a max applied to the connection; when it is reached; the client returns null.

I did set 
```
connectionFactory.setChannelCacheSize(450);
connectionFactory.setChannelCheckoutTimeout(Integer.MAX_VALUE);
```
where I would expect all threads to block and should never run out of channels but it would still generate channel max error.

This 1 channel workaround will work but it limits the rabbitmq's capability I believe.

**Comment by garyrussell** [comment]
As I said in [my comment above](https://github.com/spring-projects/spring-amqp/issues/999#issuecomment-492762750) the checkout timeout worked for me; I'll see why it doesn't work in your app.

**Comment by garyrussell** [comment]
I don't think using a single channel will make much difference, if any, at all - after all channels are multiplexed over the same connection.

**Comment by garyrussell** [comment]
>// BUG? Setting the channel cache size does not impact the client/server negotiated max channels

No; they are unrelated.

**Comment by garyrussell** [comment]
There's a bug in the checkout logic - the permit is released too early for PublisherConfirm channels.

**Comment by ldai100** [comment]
Thank you gary for all the help so far.  Much appreciated

**Comment by garyrussell** [comment]
Just for interest, with the fix in place (to block when we reach the limit), I got this result...

```
Test ran with 350005 submitted and 0 errors
Stopped HTTP test with 10 threads for minutes 1
```

Compare this to the single channel result above

```
Test ran with 1050016 submitted and 0 errors
Stopped HTTP test with 10 threads for minutes 1
```

**Comment by amitkhosla** [comment]
Hi,
Is this fix available in 2.2.1?
We are also getting the same issue with Publisher Confirm. Though our max channel count is 25.

**Comment by garyrussell** [comment]
Yes, it was merged before 2.2.0 was released. Try increasing the cache size.

If you are still having problems, start a [discussion](https://github.com/spring-projects/spring-amqp/discussions), or ask a question on Stack Overflow, showing your code, configuration, and exactly what problem you are having.
